shared/p2p/peer_manager.go

```diff
@@ -0,0 +1,200 @@
+package p2p
+
+import (
+	"sort"
+	"sync"
+
+	"github.com/pokt-network/pocket/logger"
+	"github.com/pokt-network/pocket/shared/crypto"
+)
+
+// TECHDEBT(#553): can this be simplified / consolidated in to something else?
+type PeersView interface {
+	GetAddrs() []string
+	GetPeers() PeerList
+	GetPeerstore() Peerstore
+}
+
+type PeerManager interface {
+	// TECHDEBT: move this to some new `ConnManager` interface.
+	GetSelfAddr() crypto.Address
+	GetPeersView() PeersView
+	// HandleEvent synchronously reacts to `PeerManagerEvent`s
+	HandleEvent(event PeerManagerEvent)
+}
+
+const (
+	AddPeerEventType = PeerManagerEventType(iota)
+	RemovePeerEventType
+)
+
+var _ PeerManager = &SortedPeerManager{}
+
+type PeerManagerEventType int
+
+type PeerManagerEvent struct {
+	EventType PeerManagerEventType
+	Peer      Peer
+}
+
+type SortedPeerManager struct {
+	startAddr crypto.Address
+
+	// eventCh provides a way for processing additions and removals from the pstore in an event-sourced way.
+	//
+	// The idea is that when we receive one of these events, we update the peersManager internal data structures
+	// so that it can present a consistent state but in an optimized way.
+	eventCh chan PeerManagerEvent
+
+	m  sync.RWMutex
+	wg sync.WaitGroup
+
+	sortedPeers PeerList
+	sortedAddrs []string
+	pstore      Peerstore
+}
+
+type SortedPeersView struct {
+	pstore      Peerstore
+	sortedAddrs []string
+	sortedPeers PeerList
+}
+
+// NewSortedPeersManager creates a new peersManager instance, it is in charge of
+// handling operations on sortedPeers (like adding/removing them) within a PeerAddrMap
+// it also takes care of keeping the PeerAddrMap sorted and indexed for fast access
```

Comment:
This turns out to be one of the exceptions  to lines being updated only in one commit. It looks like this gets cleaned up later to read (see [latest](https://github.com/pokt-network/pocket/pull/558/files#diff-53fb74f031fcba2237031d6006fd373fc654f152bedc6fbf778939f51bf9f10bR63)):

```
// NewSortedPeerManager creates a new SortedPeerManager instance, it is
// responsible for handling operations on `sortedAddrs` and `sortedPeers`
// (like adding/removing them) within a Peerstore. It also takes care of
// keeping them sorted and indexed for fast access.
//
// If `isDynamic` is false, the peersManager will not handle addressBook changes,
// it will only be used for querying the PeerAddrMap
// TECHDEBT: signature should include a logger reference.
```